Add DateTime/Date/TimeOfDay support. Simplify EF7 entity update logic.

rayao opened the pull request:
https://github.com/OData/RESTier/issues/273

Also, removed EF.SqlServer dependency from Restier.EF7

lewischeng-ms commented on src/Microsoft.Restier.EntityFramework/Model/ModelProducer.cs:
```diff
@@ -336,11 +336,46 @@ private static IEdmPrimitiveTypeReference GetPrimitiveTypeReference(EdmProperty
+                                return EdmCoreModel.Instance.GetDate(efProperty.Nullable);
+                            }
+                        }
+                    }
```
Is your code written specially for EF7 here? I am not sure about the difference that in my code the DB type name comes from the `ColumnAttribute`. But I guess we can keep both code here and use a compilation constant to differentiate EF6 and EF7?

rayao replied:
No this works only for EF6, EF7 has a much cleaner approach.
EF7 shares most code files from EF6, but has dedicated ModelProducer and ChangeSetPreparer.

lewischeng-ms commented on src/Microsoft.Restier.EntityFramework/Submit/ChangeSetPreparer.cs:
```diff
@@ -209,6 +209,24 @@ private static object ConvertIfNecessary(Type type, object value)
             }
 
+            // Convert to System.DateTime supported by EF from DateTimeOffset.
+            if (value is DateTimeOffset)
```
We have almost identical code here:) I just need to move some of my comments here later after your commit.

rayao replied:
If you like I can revert all EF6 changes to make you merge easier.

lewischeng-ms commented on src/Microsoft.Restier.EntityFramework7/Model/ModelProducer.cs:
```diff
@@ -202,6 +203,29 @@ private static void SetComputedAnnotation(EdmModel model, IEdmProperty target)
                 return null;
             }
+
+            if (Type.GetTypeCode(propertyType) == TypeCode.DateTime)
```
I remember some .NET profiles have retired `Type.GetTypeCode` thus we copied the whole `TypeCode` enum into ODataLib. Is there any reason we are not using `propertyType == typeof(DateTime)` here?

rayao replied:
Oh it's my fault, this is written before I brought TypeHelper in, will change it.

lewischeng-ms commented on src/Shared/TypeExtensions.cs:
```diff
@@ -105,4 +103,62 @@ private static bool IsGenericDefinition(this Type type, Type definition)
+    {
+        public static Type GetUnderlyingTypeOrSelf(Type type)
+        {
+            return Nullable.GetUnderlyingType(type) ?? type;
```
Great! I forgot to take nullable temporal types into account when comparing type in `ChangeSetPreparer` and the converters. I will change to use this helper class for other type comparisons later.

lewischeng-ms commented on src/Microsoft.Restier.Samples.Northwind/Models/NorthwindContext.cs:
```diff
@@ -121,7 +121,7 @@ protected override void OnModelCreating(Data.Entity.ModelBuilder modelBuilder)
             {
                 entityBuilder.ToTable("Orders");
-                entityBuilder.HasMany(e => e.Order_Details).WithOne(e => e.Order).Required().ForeignKey(e => e.OrderID);
+				entityBuilder.HasMany(e => e.Order_Details).WithOne(e => e.Order).Required().ForeignKey(e => e.OrderID);
```
please use space instead of tab in source code.

lewischeng-ms commented on src/Microsoft.Restier.EntityFramework7/Submit/ChangeSetPreparer.cs:
```diff
@@ -116,78 +126,110 @@ private ChangeSetPreparer()
-                // Then apply all the properties of the new instance to the instance to be updated.  This will set any unspecified
-                // properties to their default value.
+                PropertyEntry propertyEntry = dbEntry.Property(propertyPair.Key);
+                Type type = propertyEntry.CurrentValue.GetType();
```
Actually here is a bug because the current value could be null thus there is no type here. I've fixed the bug in EF6 project but not EF7. I think we shall have a task #280 to sync the latest features from EF6 to EF7.

lewischeng-ms commented on src/Microsoft.Restier.EntityFramework7/Model/ModelProducer.cs:
```diff
@@ -202,6 +203,29 @@ private static void SetComputedAnnotation(EdmModel model, IEdmProperty target)
+            if (propertyType == typeof(TimeSpan))
+            {
+                RelationalPropertyAnnotations annotations = new RelationalPropertyAnnotations(efProperty, null);
+                var columnType = annotations.ColumnType;
```
Is this new feature in EF7's `efProperty`?

rayao replied:
Yeah this relies on EF7 annotations, EF7 model configurations work in this way.

rayao commented:
Hi Lewis, perhaps I'd better remove all EF6 changes to make your merge easier? Or you can check in first and let me merge.

lewischeng-ms commented:
@rayao, my PR need to be based on yours and use some code you wrote. So please address the comments and merge this PR first if all tests can pass and there is no code analysis issue during compilation. Thanks!

rayao commented:
Sorry I messed up the branch hierarchy. Didn't notice there're check-ins before merge. Since it tells me "no conflict", I took that for "no change since last sync".
I'm not very familiar with git commands, I'm afraid of messing it up further, could someone help restore a single lineage?

lewischeng-ms commented:
@rayao Don't worry about that. You didn't mess up the master branch history. It's OK to merge a pull request like this. For future, if you want to make the commits a single line, you can first do a `git pull --rebase OData master` to pull and rebase all the latest changes from `OData/master` then do a `git push OData master` to check in.
